test/unittests/state_mpt_test.cpp

```diff
@@ -0,0 +1,333 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <gtest/gtest.h>
+#include <test/state/mpt.hpp>
+#include <test/state/rlp.hpp>
+#include <test/utils/utils.hpp>
+#include <numeric>
+
+using namespace evmone;
+using namespace evmone::state;
+using namespace intx;
+
+TEST(state_mpt, empty_trie)
+{
+    EXPECT_EQ(MPT{}.hash(), emptyMPTHash);
+}
+
+TEST(state_mpt, single_account_v1)
+{
+    // Expected value computed in go-ethereum.
+    constexpr auto expected =
+        0x084f337237951e425716a04fb0aaa74111eda9d9c61767f2497697d0a201c92e_bytes32;
+
+    constexpr auto addr = 0x0000000000000000000000000000000000000002_address;
+    constexpr uint64_t nonce = 0;
+    constexpr auto balance = 1_u256;
+    constexpr auto storage_hash = emptyMPTHash;
+    constexpr auto code_hash =
+        0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470_bytes32;
+
+    MPT trie;
+    const auto xkey = keccak256(addr);
+    auto xvalue = rlp::encode_tuple(nonce, balance, storage_hash, code_hash);
+    trie.insert(xkey, std::move(xvalue));
+    EXPECT_EQ(trie.hash(), expected);
+}
+
+TEST(state_mpt, storage_trie_v1)
+{
+    constexpr auto expected =
+        0xd9aa83255221f68fdd4931f73f8fe6ea30c191a9619b5fc60ce2914eee1e7e54_bytes32;
+
+    const auto key = 0x00_bytes32;
+    const auto value = 0x01ff_bytes32;
+    const auto xkey = keccak256(key);
+    auto xvalue = rlp::encode(rlp::trim(value));
+
+    MPT trie;
+    trie.insert(xkey, std::move(xvalue));
+    EXPECT_EQ(trie.hash(), expected);
+}
+
+TEST(state_mpt, leaf_node_example1)
+{
+    MPT trie;
+    trie.insert("\x01\x02\x03"_b, "hello"_b);
+    EXPECT_EQ(hex(trie.hash()), "82c8fd36022fbc91bd6b51580cfd941d3d9994017d59ab2e8293ae9c94c3ab6e");
+}
+
+TEST(state_mpt, branch_node_example1)
+{
+    // A trie of single branch node and two leaf nodes with paths of length 2.
+    // The branch node has leaf nodes at positions [4] and [5].
+    // {4:1, 5:a}
+
+    auto value1 = "v___________________________1"_b;
+    const auto key1 = "A"_b;
+    const uint8_t path1[]{0x4, 0x1};
+    const bytes encoded_path1{uint8_t(0x30 | path1[1])};
+    const auto leaf_node1 = rlp::encode_tuple(encoded_path1, value1);
+    EXPECT_EQ(hex(leaf_node1), "df319d765f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f5f31");
```

Comment:
well this tests just how `rlp::encode_tuple` works and doesn't influence the check for `trie.hash()` in the end...

maybe either make a separate test with these two `leaf_node` checks, or just delete them?

Reply:
This is true, they just try to reveal the internal of the trie, but these are not directly accessible by API any more. I left these as documentation for new implementations.